src/App.tsx

```diff
@@ -1,14 +1,108 @@
+import { getTodos } from './api';
+import { Todo } from './types/Todo';
+
+// function filterTodos(todos: Todo[], filterParams: FilterParamsType) {
+//   let todosCopy = [...todos];
+
+//   if (filterParams.selectFilter === FilterParams.Active && !filterParams.query) {
+//     todosCopy = todosCopy.filter((todoTask) => !todoTask.completed);
+//   }
+
+//   if (filterParams.selectFilter === FilterParams.Completed && !filterParams.query) {
+//     todosCopy = todosCopy.filter((todoTask) => todoTask.completed);
+//   }
+
+//   // if (filterParams.selectFilter === FilterParams.All && !filterParams.query) {
+//   //   todosCopy = getTodos().then();
+//   // }
+
+//   if (filterParams.query) {
+//     todosCopy = todosCopy.filter(({ title }) => title.toLowerCase()
+//       .includes(filterParams.query.toLowerCase()));
+//   }
+
+//   return todosCopy;
+// }
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState([] as Todo[]);
+
+  const [filterParam, setFilterParam] = useState({
+    selectFilter: 'all',
+    query: '',
+  });
+
+  const [userId, setUserId] = useState(0);
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>({} as Todo);
+
+  const [isLoading, setIsLoading] = useState(true);
+  const [isModalActive, setIsModalActive] = useState(false);
+
+  useEffect(() => {
+    getTodos().then(setTodos).finally(() => setIsLoading(false));
+  }, []);
+
+  const handleChangeFilterParam = (event: ChangeEvent<HTMLSelectElement>) => {
+    setFilterParam((prev) => ({ ...prev, selectFilter: event.target.value }));
+  };
+
+  const handleSearch = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setFilterParam((prev) => ({ ...prev, query: event.target.value }));
+  };
+
+  const handleResetQuery = () => {
+    setFilterParam((prev) => ({ ...prev, query: '' }));
+  };
+
+  const handleSelectUser = (id: number) => {
+    setUserId(id);
+  };
+
+  const handleSelectTodo = (todoId: number) => {
+    setSelectedTodo(todos.find(({ id }) => id === todoId) || null);
+  };
+
+  const handleHideModal = () => {
+    setSelectedTodo(null);
+    setUserId(0);
+    setIsModalActive(false);
+  };
+
+  useEffect(() => {
+    if (filterParam.selectFilter === FilterParams.All) {
+      getTodos().then(todo => {
+        setTodos(todo.filter(({ title }) => title.toLowerCase().includes(filterParam.query.toLowerCase())));
+      });
+    }
+
+    if (filterParam.selectFilter === FilterParams.Active) {
+      getTodos().then(todo => {
+        setTodos(todo.filter(({ completed, title }) => {
+          return !completed
+            && title.toLowerCase().includes(filterParam.query.toLowerCase());
+        }));
+      });
+    }
+
+    if (filterParam.selectFilter === FilterParams.Completed) {
+      getTodos().then(todo => {
+        setTodos(todo.filter(({ completed, title }) => {
+          return completed
+            && title.toLowerCase().includes(filterParam.query.toLowerCase());
+        }));
+      });
+    }
```

Comment:
create function to handle this

Reply:
Not fixed 😢

Reply:
Why do you need to do the same API call on every filter change? You already loaded data and can just filter it
Also, here you can use `switch/case`
Try to move main logic as a helper function outside the component